In [1]:
import streamlit as st
import pandas as pd
import plotly.express as px

df = pd.read_csv('vehicles_us.csv')
df.head()


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


Based on the initial observations, the data frame has 51525 entries, with a total of 13 columns. The data types seem not appropriate for few columns like 'date_posted', so I'll handle issue with the data types. Non-Null Count is not equal to the number of entries for a few columns like 'paint_color', so it's seem there is mssing values in the data. 

In [3]:
# Handle Missing Values
df.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

We have 5 columns with missing value

In [4]:
model_year_mean = df['model_year'].mean()
df.fillna({'model_year' : model_year_mean}, inplace=True)

In [5]:
# Display rows with missing values in the 'cylinders' column
missing_rows = df[df['cylinders'].isna()]
print("Rows with missing values in the 'cylinders' column:")
print(missing_rows)

Rows with missing values in the 'cylinders' column:
       price  model_year              model  condition  cylinders fuel  \
9       9200      2008.0        honda pilot  excellent        NaN  gas   
36     10499      2013.0       chrysler 300       good        NaN  gas   
37      7500      2005.0      toyota tacoma       good        NaN  gas   
59      5200      2006.0  toyota highlander       good        NaN  gas   
63     30000      1966.0       ford mustang  excellent        NaN  gas   
...      ...         ...                ...        ...        ...  ...   
51429   3250      2004.0       toyota camry       good        NaN  gas   
51442  28990      2018.0          ford f150  excellent        NaN  gas   
51460   5995      2007.0        ford fusion  excellent        NaN  gas   
51477   6499      2007.0           acura tl       good        NaN  gas   
51486   5895      2009.0   hyundai santa fe  excellent        NaN  gas   

       odometer transmission         type paint_color  is_4

In [6]:
# Calculate the median 'cylinders' for each group, and use transform to align with the original DataFrame
median_cylinder = df.groupby(['model', 'model_year'])['cylinders'].transform('median')

# Fill NaN values with the calculated median values
df['cylinders'].fillna(median_cylinder, inplace=True)

# Display the modified dataframe
print("\nModified Data Frame with NaN values replaced by median:")
print(df)


Modified Data Frame with NaN values replaced by median:
       price  model_year           model  condition  cylinders fuel  odometer  \
0       9400  2011.00000          bmw x5       good        6.0  gas  145000.0   
1      25500  2009.75047      ford f-150       good        6.0  gas   88705.0   
2       5500  2013.00000  hyundai sonata   like new        4.0  gas  110000.0   
3       1500  2003.00000      ford f-150       fair        8.0  gas       NaN   
4      14900  2017.00000    chrysler 200  excellent        4.0  gas   80903.0   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249  2013.00000   nissan maxima   like new        6.0  gas   88136.0   
51521   2700  2002.00000     honda civic    salvage        4.0  gas  181500.0   
51522   3950  2009.00000  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455  2013.00000  toyota corolla       good        4.0  gas  139573.0   
51524   6300  2014.00000   nissan altima       good 

C:\Users\luken\AppData\Local\Temp\ipykernel_39220\2492439460.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['cylinders'].fillna(median_cylinder, inplace=True)


In [7]:
odometer_mean = df['odometer'].mean()
df.fillna({'odometer' : odometer_mean}, inplace=True)

In [8]:
df.fillna({'paint_color' : 'color non available'}, inplace=True)

In [9]:
df.fillna({'is_4wd' : '99'}, inplace=True)

In [10]:
df.isna().sum()

price            0
model_year       0
model            0
condition        0
cylinders       26
fuel             0
odometer         0
transmission     0
type             0
paint_color      0
is_4wd           0
date_posted      0
days_listed      0
dtype: int64

Filled all missing value in columns model_year, cylinders and odometer by the mean of each column.
And regarding the column paint_color by no color available and  is_4wd column by 99, where the 99 will mean for me don't know.
 

In [11]:
df.fillna({'cylinders': median_cylinder}, inplace=True)

In [12]:
df.isna().sum()

price            0
model_year       0
model            0
condition        0
cylinders       26
fuel             0
odometer         0
transmission     0
type             0
paint_color      0
is_4wd           0
date_posted      0
days_listed      0
dtype: int64

In [13]:
df = df.dropna(axis=1)

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    51525 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   fuel          51525 non-null  object 
 5   odometer      51525 non-null  float64
 6   transmission  51525 non-null  object 
 7   type          51525 non-null  object 
 8   paint_color   51525 non-null  object 
 9   is_4wd        51525 non-null  object 
 10  date_posted   51525 non-null  object 
 11  days_listed   51525 non-null  int64  
dtypes: float64(2), int64(2), object(8)
memory usage: 4.7+ MB


In [15]:
#Remove Duplicates
print(df.duplicated().head())
print(df.duplicated().sum())

0    False
1    False
2    False
3    False
4    False
dtype: bool
0


Seem like there is no duplicate

In [16]:
#Correct Data Types
df['date_posted'] = pd.to_datetime(df['date_posted'])
print(df.dtypes)

price                    int64
model_year             float64
model                   object
condition               object
fuel                    object
odometer               float64
transmission            object
type                    object
paint_color             object
is_4wd                  object
date_posted     datetime64[ns]
days_listed              int64
dtype: object


In [17]:
# convert the year and odonometer to int
df['model_year'] = df['model_year'].astype(int)
df['odometer'] = df['odometer'].astype(int)
print(df.dtypes)

price                    int64
model_year               int64
model                   object
condition               object
fuel                    object
odometer                 int64
transmission            object
type                    object
paint_color             object
is_4wd                  object
date_posted     datetime64[ns]
days_listed              int64
dtype: object


In [18]:
fig = px.bar(df, x='type', y='price')
fig.show()

In [19]:
#Let define the prices category and build the plot
def price_category(row):
    if row < 3000:
        return 'price less then $5000'
    elif row >= 5000 and row <20000:
        return 'price between $5000-$20000'
    elif row >= 20000 and row <70000:
        return 'price between $20000-$70000'
    elif row >= 70000 and row <150000:
        return 'price between $70000-$150000'
    else:
        return 'over 150000'
df['price_category'] = df['price'].apply(price_category)

list_agg = df.groupby(['price_category', 'model'])['odometer'].mean().reset_index()

fig3 = px.scatter(list_agg, y='model', x='odometer', color='price_category',
             title='Number of Vehicles by Price Category and Model',
             labels={'model': 'Vehicle Model', 'odometer': 'Vehicles miles', 'price_category': 'Price Categories'})

fig3.show()

In [20]:
# distribution of vehicle prices and odometer readings
fig = px.histogram(df, x='price', title='Vehicle Price Distribution')
fig.show()

In [21]:
# Price vs. Odometer:
fig = px.scatter(df, x='odometer', y='price', color='condition',
                title='Price vs. Odometer by Vehicle Condition')
fig.show()